ratelimit/token_bucket.go

```diff
@@ -0,0 +1,67 @@
+package ratelimit
+
+import (
+	"errors"
+	"time"
+
+	"github.com/tsenart/tb"
+	"golang.org/x/net/context"
+
+	"github.com/go-kit/kit/endpoint"
+)
+
+// ErrThrottled is returned in the request path when the rate limiter is
+// triggered and the request is rejected.
+var ErrThrottled = errors.New("throttled")
+
+// NewTokenBucketThrottler returns an endpoint.Middleware that acts as a rate
+// limiter based on a "token-bucket" algorithm. Requests that would exceed the
+// maximum request rate are rejected with an error.
+func NewTokenBucketThrottler(options ...TokenBucketOption) endpoint.Middleware {
+	t := tokenBucketThrottler{
+		freq: 100 * time.Millisecond,
+		key:  "",
+		rate: 100,
+		take: 1,
+	}
+	for _, option := range options {
+		option(&t)
+	}
+	throttler := tb.NewThrottler(t.freq)
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, request interface{}) (interface{}, error) {
+			if throttler.Halt(t.key, t.take, t.rate) {
+				return nil, ErrThrottled
+			}
+			return next(ctx, request)
+		}
+	}
+}
+
+type tokenBucketThrottler struct {
+	freq time.Duration
+	key  string
+	rate int64
+	take int64
+}
+
+// TokenBucketOption sets an option on the token bucket throttler.
+type TokenBucketOption func(*tokenBucketThrottler)
```

Comment:
This is the functional options idiom. The private type is an implementation detail, users never interact with it. Still no?